firedrake/parloops.py

```diff
@@ -74,38 +72,50 @@ def _form_kernel(kernel, measure, args, **kwargs):
             # Constants modelled as Globals, so no need for double
             # indirection
             ndof = func.dat.cdim
-            kargs.append(ast.Decl("double", ast.Symbol(var, (ndof, )),
-                                  qualifiers=["const"]))
+            kargs.append(loopy.GlobalArg(var, dtype=func.dat.dtype, shape=(ndof,)))
         else:
             # Do we have a component of a mixed function?
             if isinstance(func, Indexed):
                 c, i = func.ufl_operands
                 idx = i._indices[0]._value
                 ndof = c.function_space()[idx].finat_element.space_dimension()
+                cdim = c.dat[idx].cdim
+                dtype = c.dat[idx].dtype
             else:
-                if len(func.function_space()) > 1:
-                    raise NotImplementedError("Must index mixed function in par_loop.")
-                ndof = func.function_space().finat_element.space_dimension()
+                if func.function_space().ufl_element().family() == "Real":
+                    ndof = func.function_space().dim()  # == 1
+                    kargs.append(loopy.GlobalArg(var, dtype=func.dat.dtype, shape=(ndof,)))
+                    continue
+                else:
+                    if len(func.function_space()) > 1:
+                        raise NotImplementedError("Must index mixed function in par_loop.")
+                    ndof = func.function_space().finat_element.space_dimension()
+                    cdim = func.dat.cdim
+                    dtype = func.dat.dtype
             if measure.integral_type() == 'interior_facet':
                 ndof *= 2
-            if measure is direct:
-                kargs.append(ast.Decl("double", ast.Symbol(var, (ndof,))))
-            else:
-                kargs.append(ast.Decl("double *", ast.Symbol(var, (ndof,))))
-        lkernel = lkernel.replace(var+".dofs", str(ndof))
+            # FIXME: shape for facets [2][ndof]?
+            kargs.append(loopy.GlobalArg(var, dtype=dtype, shape=(ndof, cdim)))
+        kernel_domains = kernel_domains.replace(var+".dofs", str(ndof))
 
-    body = ast.FlatBlock(lkernel)
+    if kernel_domains is "":
+        kernel_domains = "[] -> {[]}"
+    kargs.append(...)
```

Comment:
Why do we need this? I would have thought we explicitly state all kernel arguments. Or is this so that loopy can infer types of local variables?